File: java/src/main/java/com/turbotoken/ChatTemplate.java
package com.turbotoken;

import java.util.List;
import java.util.Objects;

/**
 * Chat message formatting for token counting and encoding.
 */
public final class ChatTemplate {

    /* ── ChatMessage ─────────────────────────────────────────────────── */

    public static final class ChatMessage {
        private final String role;
        private final String name;
        private final String content;

        public ChatMessage(String role, String content) {
            this(role, null, content);
        }

        public ChatMessage(String role, String name, String content) {
            this.role = Objects.requireNonNull(role, "role");
            this.name = name;
            this.content = Objects.requireNonNull(content, "content");
        }

        public String getRole()    { return role; }
        public String getName()    { return name; }
        public String getContent() { return content; }
    }

    /* ── Template ────────────────────────────────────────────────────── */

    public static final class Template {
        private final String messagePrefix;
        private final String messageSuffix;
        private final String assistantPrefix;

        public Template(String messagePrefix, String messageSuffix, String assistantPrefix) {
            this.messagePrefix = messagePrefix;
            this.messageSuffix = messageSuffix;
            this.assistantPrefix = assistantPrefix;
        }

        public String getMessagePrefix()   { return messagePrefix; }
        public String getMessageSuffix()   { return messageSuffix; }
        public String getAssistantPrefix() { return assistantPrefix; }
    }

    /* ── ChatOptions ─────────────────────────────────────────────────── */

    public static final class ChatOptions {
        private final boolean primeWithAssistantResponse;
        private final TemplateMode templateMode;

        public ChatOptions() {
            this(false, TemplateMode.TURBOTOKEN_V1);
        }

        public ChatOptions(boolean primeWithAssistantResponse, TemplateMode templateMode) {
            this.primeWithAssistantResponse = primeWithAssistantResponse;
            this.templateMode = templateMode;
        }

        public boolean isPrimeWithAssistantResponse() { return primeWithAssistantResponse; }
        public TemplateMode getTemplateMode()          { return templateMode; }
    }

    public enum TemplateMode {
        TURBOTOKEN_V1,
        IM_TOKENS
    }

    /* ── Resolve ─────────────────────────────────────────────────────── */

    /**
     * Resolves the chat template for the given mode.
     */
    public static Template resolve(TemplateMode mode) {
        switch (mode) {
            case TURBOTOKEN_V1:
                return new Template(
                    "<|im_start|>",
                    "<|im_end|>\n",
                    "<|im_start|>assistant\n"
                );
            case IM_TOKENS:
                return new Template(
                    "<|im_start|>",
                    "<|im_end|>\n",
                    "<|im_start|>assistant\n"
                );
            default:
                throw new IllegalArgumentException("Unknown template mode: " + mode);
        }
    }

    /**
     * Formats a list of chat messages into a single string for tokenization.
     */
    public static String formatMessages(List<ChatMessage> messages, ChatOptions options) {
        Template template = resolve(options.getTemplateMode());
        StringBuilder sb = new StringBuilder();

        for (ChatMessage msg : messages) {
            sb.append(template.getMessagePrefix());
            sb.append(msg.getRole());
            if (msg.getName() != null) {
                sb.append(" name=").append(msg.getName());
            }
            sb.append('\n');
            sb.append(msg.getContent());
            sb.append(template.getMessageSuffix());
        }

        if (options.isPrimeWithAssistantResponse()) {
            sb.append(template.getAssistantPrefix());
        }

        return sb.toString();
    }

    private ChatTemplate() {}
}

File: csharp/ChatTemplate.cs
using System.Collections.Generic;
using System.Text;

namespace TurboToken
{
    /// <summary>
    /// A chat message with role and content.
    /// </summary>
    public sealed class ChatMessage
    {
        public string Role { get; }
        public string? Name { get; }
        public string Content { get; }

        public ChatMessage(string role, string content, string? name = null)
        {
            Role = role;
            Content = content;
            Name = name;
        }
    }

    /// <summary>
    /// Template for formatting chat messages.
    /// </summary>
    public sealed class ChatTemplate
    {
        public string MessagePrefix { get; }
        public string MessageSuffix { get; }
        public string? AssistantPrefix { get; }

        public ChatTemplate(
            string messagePrefix = "",
            string messageSuffix = "",
            string? assistantPrefix = null)
        {
            MessagePrefix = messagePrefix;
            MessageSuffix = messageSuffix;
            AssistantPrefix = assistantPrefix;
        }
    }

    /// <summary>
    /// Chat template rendering mode.
    /// </summary>
    public enum ChatTemplateMode
    {
        TurbotokenV1,
        ImTokens,
    }

    /// <summary>
    /// Options for chat encoding.
    /// </summary>
    public sealed class ChatOptions
    {
        public bool PrimeWithAssistantResponse { get; }
        public ChatTemplate? Template { get; }
        public ChatTemplateMode Mode { get; }

        public ChatOptions(
            bool primeWithAssistantResponse = false,
            ChatTemplate? template = null,
            ChatTemplateMode mode = ChatTemplateMode.TurbotokenV1)
        {
            PrimeWithAssistantResponse = primeWithAssistantResponse;
            Template = template;
            Mode = mode;
        }
    }

    internal static class ChatFormatter
    {
        internal static string FormatChat(IReadOnlyList<ChatMessage> messages, ChatOptions options)
        {
            var sb = new StringBuilder();
            var template = options.Template ?? new ChatTemplate();

            switch (options.Mode)
            {
                case ChatTemplateMode.TurbotokenV1:
                    foreach (var msg in messages)
                    {
                        sb.Append("<|im_start|>");
                        if (msg.Name != null)
                            sb.Append($"{msg.Role} name={msg.Name}\n");
                        else
                            sb.Append($"{msg.Role}\n");
                        sb.Append($"{msg.Content}<|im_end|>\n");
                    }
                    if (options.PrimeWithAssistantResponse)
                        sb.Append("<|im_start|>assistant\n");
                    break;

                case ChatTemplateMode.ImTokens:
                    foreach (var msg in messages)
                    {
                        sb.Append(template.MessagePrefix);
                        if (msg.Name != null)
                            sb.Append($"{msg.Role} name={msg.Name}\n");
                        else
                            sb.Append($"{msg.Role}\n");
                        sb.Append(msg.Content);
                        sb.Append(template.MessageSuffix);
                    }
                    if (options.PrimeWithAssistantResponse && template.AssistantPrefix != null)
                        sb.Append(template.AssistantPrefix);
                    break;
            }

            return sb.ToString();
        }
    }
}
